returnn/tf/layers/basic.py

```diff
@@ -3686,7 +3686,8 @@ def map_axis_name(s):
         setattr(out, s, i)
         if s == "feature_dim_axis":
           out.dim = out.batch_shape[out.feature_dim_axis]
-    if size_base:
+    if out.size_placeholder is not None and size_base:  # size_placeholder is not set if source is DataNotAvailableLayer
```

Comment:
I don't quite understand this comment. Why is this only for DataNotAvailableLayer, and maybe not in other cases? And why is that way at all? Shouldn't we maybe fix this in DataNotAvailableLayer? This sounds like you do some workaround here.

Reply:
I think the `size_placeholder` being `None` is a valid case. This can not be fixed in the DataNotAvailableLayer, because the (correct) content for the size_placeholder does not exist, and the code would fail in any way. Introducing an arbitrary filled size_placeholder does not sound like a valid solution.

Reply:
But then this comment is wrong (misleading), as this is not about DataNotAvailableLayer.

Reply:
Also, as I always say, we should never just accept it as it is. We should always ask: Does this makes sense? Should it really be this way? Does it make sense that `size_placeholder` can be `None`?

Reply:
But it does make sense. What should the `size_placeholder` be if the data does not exist?  I think `None` is valid here. Of course you could add a `size_placeholder` with fake entries, but I do not think this is a good idea. Any variable should have a reasonable content, or not?

The comment is also correct, because as long as we are not using `DataNotAvailableLayer`, the size_placeholder should be correctly defined.

Reply:
But there might be other layers as well. It would be wrong to say this is only for one specific layer. The `LayerBase` class is the base and should never have specific code for specific sub instances.

Either we say `size_placeholder` can be `None` in general, or not. But saying it can only be for one specific layer is very hacky and will surely lead to other problems.

What you could do (much better):
```suggestion
    if out.size_placeholder is not None and size_base:  # size_placeholder could be unset eg. with DataNotAvailableLayer
```